File: Beecrowd/EtiquetasDeNoel/Noel.java
/******************************************************************************
 * 
 *  ETIQUETAS DE NOEL - BEECROWD
 *  
 *  Autor: Theo Diniz Viana
 *  Nota: 100/100
 *  Linguagem: JAVA 19
 *  Data: 22/03/2025
 ******************************************************************************/
import java.util.*;
public class Noel{
    public static void compara(Destinatarios[] destinos, Carta[] cartas){
        for(int i = 0; i < destinos.length; i++){
            String lingua = destinos[i].idioma;
            for(int j = 0; j < cartas.length; j++){
                if(lingua.equals(cartas[j].idioma)){
                    System.out.println(destinos[i].nome + "\n" + cartas[j].mensagem + "\n");
                }
            }
        }
    }
    public static void main(String[] args) {
        Scanner rc = new Scanner(System.in);
        int qntCartas = rc.nextInt();
        rc.nextLine();
        Carta[] cartas = new Carta[qntCartas];

        for(int i = 0; i < qntCartas; i++){
            String idioma = rc.nextLine();
            String mensagem = rc.nextLine();
            cartas[i] = new Carta(idioma, mensagem);
        }
        int qntDestinatarios = rc.nextInt();
        rc.nextLine();

        Destinatarios[] destinos = new Destinatarios[qntDestinatarios];

        
        for(int j = 0; j < qntDestinatarios; j++){
            String nome = rc.nextLine();
            String idioma = rc.nextLine();
            destinos[j] = new Destinatarios(idioma, nome);
        }

        compara(destinos, cartas);

    }
}

class Carta{
    String idioma;
    String mensagem;

    public Carta(String idioma, String mensagem){
        this.idioma = idioma;
        this.mensagem = mensagem;
    }
}

class Destinatarios{
    String idioma;
    String nome;
    
    public Destinatarios(String idioma, String nome){
        this.idioma = idioma;
        this.nome = nome;
    }
}

File: Beecrowd/EtiquetasDeNoel/noel.c
/******************************************************************************
 * 
 *  ETIQUETAS DE NOEL - BEECROWD
 *  
 *  Autor: Theo Diniz Viana
 *  Nota: 100/100
 *  Linguagem: C99
 *  Data: 21/03/2025
 ******************************************************************************/
#include <stdio.h>
#include <stdlib.h>

typedef struct
{
    char *idioma;
    char *mensagem;
} Carta;

typedef struct
{
    char *idioma;
    char *nome;
} Destinatario;

Carta *recebeCarta()
{
    Carta *carta = malloc(sizeof(Carta));
    carta->idioma = malloc(sizeof(char) * 100);
    carta->mensagem = malloc(sizeof(char) * 100);

    scanf((" %[^\n]"), carta->idioma);
    scanf((" %[^\n]"), carta->mensagem);
    return carta;
}

Destinatario *recebeDestinatario()
{
    Destinatario *destino = malloc(sizeof(Destinatario));
    destino->idioma = malloc(sizeof(char) * 100);
    destino->nome = malloc(sizeof(char) * 100);

    scanf((" %[^\n]"), destino->nome);
    scanf((" %[^\n]"), destino->idioma);
    return destino;
}

void comparar(Destinatario **destinos, Carta **cartas, int qntCartas, int qntDestinos)
{
    for (int j = 0; j < qntDestinos; j++)
    {
        char *lingua = destinos[j]->idioma;
        for (int i = 0; i < qntCartas; i++)
        {
            if (strcmp(cartas[i]->idioma, lingua) == 0)
            {
                printf("%s\n%s\n", destinos[j]->nome, cartas[i]->mensagem);
            }
        }
    }
}

int main()
{
    int quantLinguas = 0;
    scanf(" %d", &quantLinguas);
    Carta **cartas = malloc(quantLinguas * sizeof(Carta *));
    for (int j = 0; j < quantLinguas; j++)
    {
        cartas[j] = recebeCarta();
    }

    int destinatarios = 0;
    scanf(" %d", &destinatarios);

    // Alocando memoria para destinatarios
    Destinatario **destinos = malloc(destinatarios * sizeof(Destinatario *));
    for (int f = 0; f < destinatarios; f++)
    {
        destinos[f] = recebeDestinatario();
    }

    comparar(destinos, cartas, quantLinguas, destinatarios);

    // Libera memoria
    for (int l = 0; l < quantLinguas; l++)
    {
        free(cartas[l]->idioma);
        free(cartas[l]->mensagem);
        free(cartas[l]);
    }
    free(cartas);

    for (int m = 0; m < destinatarios; m++)
    {
        free(destinos[m]->idioma);
        free(destinos[m]->nome);
        free(destinos[m]);
    }
    free(destinos);
}